### ISSUE TITLE
Addition of Validated entity state

### ISSUE THREAD

**Issue by NSUWAL123** [hotosm/fmtm#2011]
**Is your feature request related to a problem? Please describe.**
We have 4 statuses for an entity:
   ```
 READY = 0
    OPENED_IN_ODK = 1
    SURVEY_SUBMITTED = 2
    MARKED_BAD = 6
```

Currently, when a submission is marked bad, the entity status is to be updated to MARKED_BAD. There's no VALIDATED state. The SURVEY_SUBMITTED case is only ideal if there's no BAD entity.
But when a submission is made again to the BAD entity and the submission is approved, what should the entity status be?  `1.` Fallback to SURVEY_SUBMITTED which doesn't make sense, or `2.` VALIDATED. 

**Describe the solution you'd like**
So, to solve the problem we should add a new status called VALIDATED.

**Comment by spwoodcock** [comment]
SURVEY_SUBMITTED is essentially an indicator to say the mapping has been done.

The validation status is handled elsewhere: internally in the ODK database, so we can access the validation state from the ODK API (avoiding the need to store this twice).

The BAD status was added here to indicate more easily the features that failed validation, without having to call the ODK API directly (its a small amount of necessary duplication).

But when the user submits another submission to override / correct the previous one, this submission will have to be validated.

It makes sense to me to mark the status as SURVEY_SUBMITTED once again, then either leave it as it is, or mark BAD if it fails validation again. 

Does that make sense?

**Comment by Sujanadh** [comment]
I think we need a status for validated features; from a validator perspective, I wouldn't know which feature is validated since every feature would have status `survey_submitted` whether they are just submitted or have been validated. If there is a distinct status, then I would know which feature to pick during validation. We only have validation status, which is `approved` in individual submissions. If validators approve the submission, then we can update the status of the entity to validated. When all features are validated within a task, then we can auto-change the task status to validated.

**Comment by spwoodcock** [comment]
But there is a validation field built into ODK right?

It seems like duplication to ignore that and put the validation in FMTM instead.

Plus validation should apply to submissions for features, not to the features themselves.

You mention a workflow where we 'pick features' for validation. But instead I think we should base it on if there are submissions. If we have an unreviewed submission, then we prompt for it to be reviewed, then marked as validated in ODK

**Comment by spwoodcock** [comment]
Essentially what I'm trying to convey is that all the 'feature' level statuses were designed around mapping.

The only time validation is involved is for marking the feature bad for the mapper to see, then re-map.

The actual validation state for submissions of features is stored in ODK

**Comment by NSUWAL123** [comment]
You are right in the context of the mapper frontend. But when the validator validates the entity via the project page through the map, how does the validator know that they have validated the feature? 
For example, thousands of entities could be out there in a single task. Suppose the `survey_submitted` and `validated` status would be specified with green color then wouldn't it be hard for the validator navigating through thousands of entities to find the next entity to be validated as all entity colors would be same?

**Comment by spwoodcock** [comment]
Yeah you make a good point!

This all depends on if we want to store the feature level validation state in ODK, or in FMTM.

ODK manages the validation status per submission, while we would manage it per entity in FMTM

**Comment by manjitapandey** [comment]
with #1946, Since we are changing the validation workflow to directly go to validation page from project details page. It will be essential for validators to know which features are already validated and which are at ready to validation state. 

So we need to show different colour for validated and ready to validate features in project details page. Can we obtain that using feature level validation state in ODK ??

**Comment by spwoodcock** [comment]
Depends on if there is an API to get the review state of all submissions on a project - not at my computer to check.

But the fallback is to add a new state for entities as everyone has mentioned. The main downside there is that we can handle multiple submissions, as we will do entity based validation instead of submission based validation

**Comment by Sujanadh** [comment]
This issue is all related to map view. Retrieving all review states will also need to have entity id in the response. There is n't any api available to get just review states of all submissions. We can get them from the api to fetch all submissions metadata only but there won't be any entities id linked. This approach will fall back to getting all submissions data , get review states and entity ids which itself  is very slow process and complicated.

**Comment by spwoodcock** [comment]
Thanks for the info 👍

So just to clarify, we are suggesting:

- Add a VALIDATED status for the features / entities in our db.

- Still use the ODK review state for **submissions**.

- This way its possible to review multiple submissions for an entity, however the entity status will become VALIDATED aften **one** submission review is done.

Sm I correct so far?

## Question

- The odk_entities table was made to make the entity statuses more easily available in the frontend via electric.

- However, now we also have a table to track 'new' geometries (and 'bad' geometries, but these are duplicated).

How would we manage the VALIDATED state of the new geometries / entities?

**Comment by Sujanadh** [comment]
Thanks for the detail breakdown.

You are correct, we use review state of submission for validation, and make the feature validated if any submission in case of multiple submissions is approved.

Regarding your question on how we manage validated state for the new geometries.
We will create entity once geometry is drawn, so that it is loaded in odk collect. They are also saved in the geometrylog table, as far as I know the purpose of creating geometrylog table for bad geom and new geom is to provide instant pulse effect and visualization in outmost layer in map via electric sql. Without table, we need to load the entire fgb and entities status to check for the bad geom. Once the new geom is created as an entity in odk, the workflow for the validation will be same as it is for all the geom. When user clicks new geom or bad geom we load that tasks entities and allow them to validate it. 

In case of bad geom, the entry will be duplicated, there will be record in both table odk_entities and geometrylog. I like the idea you proposed to add geojson column in odk_entities, in case of new geom, the question is when user draws a polygon should it be visible instantly to all the users? if not then we can create entity and load it by syncing status and use electric sql to fetch new geom or bad geom only, which might not be real time. We also need to add new status "NEW" as entity status as well. This approach will help in validation process we don't have to load tasks entities, user can just click the new feature or bad feature and start validation.

**Comment by spwoodcock** [comment]
The actionable item here is adding two additional enum states to `EntityState` (both backend and frontend):
- [ ] NEW_GEOM = 3
- [ ] VALIDATED = 5

> It might be nice to remove these numeric values in future too, simply setting NEW_GEOM = "NEW_GEOM", but that's a problem for another day too!

**Comment by manjitapandey** [comment]
@NSUWAL123 please update the feature colour to validated on project details page for the already existing entities. 
For new entities blocked by : #1979
